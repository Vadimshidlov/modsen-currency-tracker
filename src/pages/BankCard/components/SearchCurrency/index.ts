export { default } from "@/pages/BankCard/components/SearchCurrency/SearchCurrency";
