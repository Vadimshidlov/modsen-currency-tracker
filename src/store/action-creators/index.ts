export { getMapData } from "@/store/action-creators/getMapData";
export { getCurrency } from "@/store/action-creators/getCurrency";
export {
    selectSecondCurrency,
    closeModalWindow,
    openModalWindow,
} from "@/store/action-creators/openModalWindow";
export { switchLightTheme, switchDarkTheme } from "@/store/action-creators/toggleTheme";
export {
    errorAuthUserLocation,
    successAuthUserLocation,
} from "@/store/action-creators/authUserLocation";
