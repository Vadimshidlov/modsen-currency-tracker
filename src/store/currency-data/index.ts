export { mockMapData } from "@/store/currency-data/mockMapData";
export { mockBanksApiData } from "@/store/currency-data/mockBanksApiData";
export { mockStockData } from "@/store/currency-data/mockStockData";
