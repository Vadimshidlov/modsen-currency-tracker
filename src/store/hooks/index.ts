export { useActions } from "@/store/hooks/useActions";
export { useTypedSelectorHook } from "@/store/hooks/useTypedSelector";
