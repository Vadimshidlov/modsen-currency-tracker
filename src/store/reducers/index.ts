import { combineReducers } from "redux";
import themeReducer from "@/store/reducers/themeReducer";
import latestCurrencyReducer from "@/store/reducers/latestCurrencyReducer";
import modalWindowReducer from "@/store/reducers/modalWindowReducer";
import mapDataReducer from "@/store/reducers/mapDataReducer";

export const rootReducer = combineReducers({
    theme: themeReducer,
    currency: latestCurrencyReducer,
    modalWindow: modalWindowReducer,
    mapData: mapDataReducer,
});

export type RootStateType = ReturnType<typeof rootReducer>;
