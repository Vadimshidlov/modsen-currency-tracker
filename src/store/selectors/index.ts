export { selectCurrency, selectModal, selectTheme } from "@/store/selectors/selectors";
