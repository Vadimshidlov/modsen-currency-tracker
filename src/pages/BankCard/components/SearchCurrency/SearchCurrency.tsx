import React, { Component } from "react";
import { connect } from "react-redux";
import { RootStateType } from "@/store/reducers";
import TextInput from "@/pages/BankCard/components/TextInput/index";
import { SearchCurrencyPropsType, SearchCurrencyStateType } from "@/types/BankCardPageTypes";

class SearchCurrency extends Component<SearchCurrencyPropsType, SearchCurrencyStateType> {
    render() {
        const { handleSelectCurrency, currencyData, searchCurrencyValue, onSearchCurrencyValue } =
            this.props;

        const filteredCurrencyKeys = ["BYN", ...Object.keys(currencyData.currency.data)].filter(
            (key) => key.toLowerCase().includes(searchCurrencyValue.toLowerCase()),
        );

        return (
            <div className="currency-input__container">
                <TextInput
                    id="currency-input"
                    className="currency-input__input"
                    placeholder="Ð¡urrency search..."
                    value={searchCurrencyValue}
                    onChange={onSearchCurrencyValue}
                    autoComplete="off"
                    data-testid="map-search-input"
                />
                <div className="currency-input__result">
                    {searchCurrencyValue &&
                        filteredCurrencyKeys.map((currencyKey) => (
                            <button
                                className="currency-input__result-item"
                                onClick={() => {
                                    handleSelectCurrency(currencyKey);
                                }}
                                key={currencyKey}
                            >
                                {currencyKey}
                            </button>
                        ))}
                    {filteredCurrencyKeys.length === 0 && (
                        <button
                            className="currency-input__result-item"
                            key={String(filteredCurrencyKeys.length)}
                        >
                            Not Found
                        </button>
                    )}
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state: RootStateType) => ({
    currencyData: state.currency,
});

export default connect(mapStateToProps, {})(SearchCurrency);
