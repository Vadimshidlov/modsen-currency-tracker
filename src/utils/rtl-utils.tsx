import React, { PropsWithChildren } from "react";
import { Store, applyMiddleware, createStore } from "redux";
import { thunk } from "redux-thunk";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { Provider } from "react-redux";
import { RootStateType, rootReducer } from "@/store/reducers";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
    preloadedState?: Partial<RootStateType>;
    store?: Store;
}

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        store = createStore(rootReducer, preloadedState, applyMiddleware(thunk)),
        ...renderOptions
    }: ExtendedRenderOptions = {},
) {
    function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
        return <Provider store={store}>{children}</Provider>;
    }
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from "@testing-library/react";

export { renderWithProviders as render };
