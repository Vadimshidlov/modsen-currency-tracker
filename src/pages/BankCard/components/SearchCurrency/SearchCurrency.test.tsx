import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import { fireEvent, waitFor } from "@testing-library/react";
import SearchCurrency from "@/pages/BankCard/components/SearchCurrency/SearchCurrency";
import { store } from "@/store";
import { renderWithProviders, screen } from "@/utils/rtl-utils";

afterAll(() => {
    jest.clearAllMocks();
});

beforeEach(() => {
    jest.clearAllMocks();
});

const mockResponse = {
    meta: {
        last_updated_at: "2024-03-04T23:59:59Z",
    },
    data: {
        USD: { code: "USD", value: 1 },
        EUR: { code: "EUR", value: 1 },
        BTC: { code: "BTC", value: 1 },
        JPY: { code: "JPY", value: 1 },
        AUD: { code: "AUD", value: 1 },
    },
};

describe("Footer component tests", () => {
    it("should match snapshot", () => {
        const onSearchCurrencyValue = jest.fn();
        const handleSelectCurrency = jest.fn();

        const tree = renderer
            .create(
                <Provider store={store}>
                    <Router>
                        <SearchCurrency
                            onSearchCurrencyValue={onSearchCurrencyValue}
                            searchCurrencyValue=""
                            handleSelectCurrency={handleSelectCurrency}
                        />
                    </Router>
                </Provider>,
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it("should switch theme when toggle is clicked", async () => {
        const onSearchCurrencyValue = jest.fn();
        const handleSelectCurrency = jest.fn();

        renderWithProviders(
            <Router>
                <SearchCurrency
                    onSearchCurrencyValue={onSearchCurrencyValue}
                    searchCurrencyValue=""
                    handleSelectCurrency={handleSelectCurrency}
                />
            </Router>,
            {
                preloadedState: {
                    currency: {
                        currency: mockResponse,
                        isLoading: false,
                        error: null,
                    },
                    modalWindow: {
                        isOpen: false,
                        currentCurrencyTitle: "",
                        currentCurrencyCode: "",
                        currentCurrencyValue: 0,
                        secondCurrencyTitle: "",
                        secondCurrencyCode: "",
                        secondCurrencyValue: 0,
                    },
                },
            },
        );

        await waitFor(() => {
            expect(screen.getByTestId("map-search-input")).toBeInTheDocument();

            const input = screen.getByTestId("map-search-input");

            fireEvent.change(input, { target: { value: "USD" } });

            expect(onSearchCurrencyValue).toHaveBeenCalledTimes(3);
        });
    });
});

describe("shuold print and renred correct data", () => {
    it("should", async () => {
        const onSearchCurrencyValue = jest.fn();
        const handleSelectCurrency = jest.fn();

        renderWithProviders(
            <Router>
                <SearchCurrency
                    onSearchCurrencyValue={onSearchCurrencyValue}
                    searchCurrencyValue="u"
                    handleSelectCurrency={handleSelectCurrency}
                />
            </Router>,
            {
                preloadedState: {
                    currency: {
                        currency: mockResponse,
                        isLoading: false,
                        error: null,
                    },
                    modalWindow: {
                        isOpen: false,
                        currentCurrencyTitle: "",
                        currentCurrencyCode: "",
                        currentCurrencyValue: 0,
                        secondCurrencyTitle: "",
                        secondCurrencyCode: "",
                        secondCurrencyValue: 0,
                    },
                },
            },
        );

        await waitFor(() => {
            expect(screen.getByTestId("map-search-input")).toBeInTheDocument();

            expect(screen.getByText("USD")).toBeInTheDocument();
            expect(screen.getByText("EUR")).toBeInTheDocument();
            expect(screen.getByText("AUD")).toBeInTheDocument();
        });
    });
});
