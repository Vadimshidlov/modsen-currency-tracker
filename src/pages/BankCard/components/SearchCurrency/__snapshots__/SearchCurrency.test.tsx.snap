// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component tests should match snapshot 1`] = `
<div
  className="currency-input__container"
>
  <input
    autoComplete="off"
    className="currency-input__input"
    data-testid="map-search-input"
    id="currency-input"
    onChange={[MockFunction]}
    placeholder="Ð¡urrency search..."
    type="text"
    value=""
  />
  <div
    className="currency-input__result"
  />
</div>
`;
